### PR TITLE
Add OASIS-coupling for CLM3.5-ParFlow-PDAF

### PR DESCRIPTION
**Pull request by jjokella**
This is a follow-up on https://github.com/HPSCTerrSys/TSMP2/pull/39 

In #39, CLM3.5-PDAF / CLM3.5-ParFlow-PDAF could be compiled.

However, there were errors in executing the CLM3.5-ParFlow-PDAF Testcase `clm3_pfl_pdaf_fsday4` related to the OASIS-coupler and in particular the fact that the CLM3.5-ParFlow coupling is performed with an older version of the coupler.

I try to tackle the issue with this PR. My goal: That I can perform my PDAF-testcase-suite completely in the TSMP2 framework.

This is a big step towards moving all PDAF-related users from TSMP1 to TSMP2. 

## Changes

These changes only work in combination with OASIS3-MCT-v2.0, pre-patched for PDAF from the branch https://icg4geo.icg.kfa-juelich.de/ModelSystems/tsmp_src/oasis3-mct/-/tree/oasis3-MCT_2.0-pdaf-patched 

- Add submodule `oasis3-mct-clm35` for software OASIS3-MCT-v2 that is patched for CLM3.5-ParFlow-PDAF
- `build_tsmp2.sh`: Small fix in `dwn_compsrc()`, regarding the condition, when to download ParFlow from `parflow_pdaf` or OASIS3-MCT from `oasis3-mct-clm35`
- `BuildCLM3.5.cmake`: Add include directory for OASIS3-MCT-v2 modules
- `BuildOASIS3MCT.cmake`: Add completely remodelled `make.inc` for OASIS3-MCT-v2 (modelled after https://github.com/HPSCTerrSys/TSMP/blob/master/bldsva/intf_oas3/oasis3-mct/arch/config/make.intel_oa3 )
- `BuildOASIS3MCT.cmake`: Adapting `make`-target, when `CLM3.5` is used

### CODE REVIEW

**Comment by jjokella** [comment]
Testcases are now reproducing results from TSMP1 :+1: 

I think this can be merged.

Small thing bugging me: The old OASIS needs password authentication for the clone. However, I curently do not see a way around it.

**Comment by kvrigor** [comment]
I  believe CLM3.5-ParFlow should compile with OASI3-MCT v5; there's specific commits in [CLM3.5](https://github.com/HPSCTerrSys/CLM3.5/commit/a5b11a54064b8e9890dcd189f8d7bb247f6093c9) and [ParFlow](https://github.com/parflow/parflow/commit/b13945d80fda5f76c0bb7d457dda3586f5e776a7) to support this. Also we have these test cases:

- [NRW_ParFlow-CLM3.5](https://icg4geo.icg.kfa-juelich.de/Configurations/CLM/NRW_ParFlow-CLM3.5)
- [EU11_ParFlow-CLM3.5](https://icg4geo.icg.kfa-juelich.de/Configurations/CLM/eu11_parflow-clm3.5)

@jjokella can you try compiling with the latest OASIS3-MCT? If there are compile errors I think they should be easily fixable.

File: cmake/BuildOASIS3MCT.cmake [hunk 1]
```diff
@@ -16,45 +16,132 @@ set(OASIS_BLD_DIR ${CMAKE_CURRENT_BINARY_DIR}/OASIS3_MCT/bld)
...
-file(APPEND  ${OASIS_MAKE_INC} "LD              = ${CMAKE_Fortran_COMPILER} -L${MPI_Fortran_LIB_DIR}\n")
-file(APPEND  ${OASIS_MAKE_INC} "AR              = ar\n")
-file(APPEND  ${OASIS_MAKE_INC} "ARFLAGS         = -ruv\n")
-file(APPEND  ${OASIS_MAKE_INC} "DYNOPT          = -fPIC\n")
-file(APPEND  ${OASIS_MAKE_INC} "LDDYNOPT        = -shared\n")
-if(${PDAF})
-  file(APPEND  ${OASIS_MAKE_INC} "CPPDEF          = -Duse_netCDF -Duse_comm_$(CHAN) -D__VERBOSE  -DTREAT_OVERLAY -DUSE_PDAF\n")
+
+
+if(${CLM3.5})
```
Comment (kvrigor) [inline cmake/BuildOASIS3MCT.cmake:21]: You can try to disable this when compiling with OASIS3-MCT v5. I think the default `make.inc` should work for all component models.

**Comment by jjokella** [comment]
Hey @kvrigor  thanks for the review.

Actually compilation worked with CLM3.5-ParFlow-PDAF with  OASI3-MCT v5 also before this PR. There were runtime errors in the PDAF-testcase [clm3_pfl_pdaf_fsday4](https://icg4geo.icg.kfa-juelich.de/ExternalRepos/tsmp-pdaf/TSMP-PDAF-testcases/-/tree/master/clm3_pfl_pdaf_fsday4?ref_type=heads).

As I think we should not invest too much time into CLM3.5, what I could do is to put all conditions, where I change things for CLM3.5 to actually only change for CLM3.5+PDAF. Then my testcases run and non-DA CLM3.5-ParFlow will be compiled as before!? I did not realize there were still testcases for CLM3.5-PDAF that I would possibly interfere with!


### FYI: Runtime error of `clm3_pfl_pdaf_fsday4` with OASIS-MCT_v5

standard error
```
 (oasis_abort) ABORT: CALLING ABORT FROM OASIS LAYER NOW
 (oasis_abort) ABORT: See the log files in the run directory
application called MPI_Abort(comm=0xC4000000, 1) - process 0
p_Abort(r188): aborting on users request
```

oasis log file
```
(oasis_init_comp) OASIS RUNNING 
(oasis_init_comp) OPEN debug file for pe, unit :       0    9999
 (check_status) ERROR:  netcdf status
 (oasis_io_write_fldattr) nf90_strerror = NetCDF: Unknown file format
 (oasis_abort) ABORT: file      = mod_oasis_io.F90
 (oasis_abort) ABORT: line      =          266
 (oasis_abort) ABORT: on model  = oaspfl
 (oasis_abort) ABORT: on global rank =            0
 (oasis_abort) ABORT: on local  rank =            0
 (oasis_abort) ABORT: CALLING ABORT FROM OASIS LAYER NOW
```

**Comment by jjokella** [comment]
Closing this PR as CLM3.5-ParFlow-PDAF was successfully run with OASIS3-MCT-v5 after adding `$NCDFTYP` input in `namcouple`. See https://icg4geo.icg.kfa-juelich.de/ExternalRepos/tsmp-pdaf/TSMP-PDAF-testcases/-/commit/91578ef195d6cc04c6497bae2ed8b7caf396bc22

This way, this PR boils down to the small fix presented in #41

**Pull request closed without merging**